rust/osm2lanes/src/transform/error.rs

```diff
@@ -0,0 +1,198 @@
+        }
+    }
+    pub fn unsupported_str(description: &str) -> Self {
+        Self::Unsupported {
+            description: Some(description.to_owned()),
+            tags: None,
+        }
+    }
+}
+
+impl std::fmt::Display for RoadMsg {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::Deprecated {
+                deprecated_tags, ..
+            } => write!(
+                f,
+                "deprecated: {}",
+                deprecated_tags.to_vec().as_slice().join(" ")
+            ),
+            Self::Unsupported { description, tags }
+            | Self::Unimplemented { description, tags }
+            | Self::Ambiguous { description, tags } => {
+                let tags = tags.as_ref().map(|tags| tags.to_vec().as_slice().join(" "));
+                let prefix = match self {
+                    Self::Unsupported { .. } => "unsupported",
+                    Self::Unimplemented { .. } => "unimplemented",
+                    Self::Ambiguous { .. } => "ambiguous",
+                    _ => unreachable!(),
+                };
+                match (description, tags) {
+                    (None, None) => write!(f, "{}", prefix),
+                    (Some(description), None) => {
+                        write!(f, "{}: {}", prefix, description)
+                    }
+                    (None, Some(tags)) => write!(f, "{}: {}", prefix, tags),
+                    (Some(description), Some(tags)) => {
+                        write!(f, "{}: {}, {}", prefix, description, tags)
+                    }
+                }
+            }
+            Self::Other { description, .. } => write!(f, "{}", description),
+            Self::TagsDuplicateKey(e) => e.fmt(f),
+        }
+    }
+}
+
+fn use_display<T, S>(value: &T, serializer: S) -> Result<S::Ok, S::Error>
+where
+    T: std::fmt::Display,
+    S: serde::Serializer,
+{
+    serializer.collect_str(value)
+}
+
+impl Serialize for RoadMsg {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.collect_str(self)
+    }
+}
+
+#[derive(Clone, Debug, Default, Serialize)]
+pub struct RoadWarnings(Vec<RoadMsg>);
+
+impl RoadWarnings {
+    pub fn new(msgs: Vec<RoadMsg>) -> Self {
+        Self(msgs)
+    }
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+    pub fn push(&mut self, msg: RoadMsg) {
+        self.0.push(msg)
+    }
+}
+
+impl std::fmt::Display for RoadWarnings {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            self.0
+                .iter()
+                .map(|warning| format!("Warning: {}", warning))
+                .collect::<Vec<_>>()
+                .join("\n")
+        )
+    }
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub enum RoadError {
+    #[serde(serialize_with = "use_display")]
```

Comment:
Is this redundant? There's an explicit `impl Serialize for RoadMsg`.

... I got curious, and it does seem redundant. I'd suggest adding a doc test above to show the JSON serialization:

```
/// let err = RoadError::from(RoadMsg::deprecated_tag("foo", "bar"));
/// assert_eq!("{\"Msg\":\"unsupported: foo=bar\"}", serde_json::to_string(&err).unwrap());
```